distributed/diagnostics/tracer.py

```diff
@@ -0,0 +1,115 @@
+from __future__ import print_function, division, absolute_import
+
+import logging
+import json
+import time
+from tornado import gen
+
+from .plugin import SchedulerPlugin
+from ..utils import key_split
+
+
+logger = logging.getLogger(__name__)
+
+
+class TraceLoggerPlugin(SchedulerPlugin):
+    def __init__(self, scheduler):
+        self.log = None
+        logger.debug('Set up Scheduler trace logger')
+        self.scheduler = scheduler
+        self.scheduler.handlers['start_trace'] = self.start_trace
+        self.scheduler.handlers['stop_trace'] = self.stop_trace
+
+    def start_trace(self, *args, **kwargs):
+        log_file = kwargs.pop('log_file', 'trace.log')
+        logger.debug('Start trace to file: ' + log_file)
+        log = logging.getLogger('scheduler_trace')
+        json_handler = logging.handlers.RotatingFileHandler(filename=log_file)
+        json_handler.setFormatter(JSONFormatter())
+        log.addHandler(json_handler)
+        log.setLevel(logging.DEBUG)
+        self.log = log
+        self.log.debug({'log': 'start'})
+        self.scheduler.add_plugin(self)
+
+    def stop_trace(self, *args, **kwargs):
+        logger.debug('Stop trace')
+        self.log.debug({'log': 'stop'})
+        self.scheduler.extensions.pop('trace', None)
+        self.log = None
+
+    def transition(self, key, start, finish, *args, **kwargs):
+        if self.log:
+            kwargs.pop('type', None)
+            self.log.debug({'key': key, 'start': start, 'finish': finish,
+                           'args': args, 'kwargs': kwargs, 'log': 'transition'})
```

Comment:
There are many transitions here that you may not want to record.  I typically filter here for transitions that leave the task in a finished state (memory, error)